lib/registry/parsers.ts

```diff
@@ -0,0 +1,377 @@
+	let repoNameFromArg: string;
+	const protoSepIdx = arg.indexOf('://');
+	let index: ParsedIndex;
+
+	if (protoSepIdx !== -1) {
+		// (A) repo with a protocol, e.g. 'https://host/repo'.
+		const slashIdx = arg.indexOf('/', protoSepIdx + 3);
+		if (slashIdx === -1) {
+			throw new Error(
+				'invalid repository name, no "/REPO" after ' + 'hostame: ' + arg,
+			);
+		}
+		const indexName = arg.slice(0, slashIdx);
+		repoNameFromArg = arg.slice(slashIdx + 1);
+		index = parseIndex(indexName);
+	} else {
+		const parts = splitStrOnce(arg, '/');
+		if (
+			parts.length === 1 ||
+			/* or if parts[0] doesn't look like a hostname or IP */
+			(parts[0].indexOf('.') === -1 &&
+				parts[0].indexOf(':') === -1 &&
+				parts[0] !== 'localhost')
+		) {
+			// (B) repo without leading 'INDEX/'.
+			index = parseIndex(undefined);
+			repoNameFromArg = arg;
+		} else {
+			// (C) repo with leading 'INDEX/' (without protocol).
+			index = parseIndex(parts[0]);
+			repoNameFromArg = parts[1];
+		}
+	}
+
+	const nameParts = splitStrOnce(repoNameFromArg, '/');
+	let ns: string | undefined;
+	let name: string;
+	if (nameParts.length === 2) {
+		name = nameParts[1];
+
+		// Validate ns.
+		ns = nameParts[0];
+		if (ns.length < 2 || ns.length > 255) {
+			throw new Error(
+				'invalid repository namespace, must be between ' +
+					'2 and 255 characters: ' +
+					ns,
+			);
+		}
+		if (!VALID_NS.test(ns)) {
+			throw new Error(
+				'invalid repository namespace, may only contain ' +
+					'[a-z0-9._-] characters: ' +
+					ns,
+			);
+		}
+		if (ns[0] === '-' && ns[ns.length - 1] === '-') {
+			throw new Error(
+				'invalid repository namespace, cannot start or end with a hypen: ' + ns,
+			);
+		}
+		if (ns.indexOf('--') !== -1) {
+			throw new Error(
+				'invalid repository namespace, cannot contain ' +
+					'consecutive hyphens: ' +
+					ns,
+			);
+		}
+	} else {
+		name = repoNameFromArg;
+		if (index.official) {
+			ns = 'library';
+		}
+	}
+
+	// Validate name.
+	if (!VALID_REPO.test(name)) {
+		throw new Error(
+			'invalid repository name, may only contain [a-z0-9_/.-] characters: ' +
+				name,
+		);
+	}
+
+	if (index.official) {
+		if (!ns) {
+			throw new Error('Namespace undetermined');
+		}
+		info.remoteName = ns + '/' + name;
+		if (ns === 'library') {
+			info.localName = name;
+		} else {
+			info.localName = info.remoteName;
+		}
+		info.canonicalName = DEFAULT_INDEX_NAME + '/' + info.localName;
+		info.indexUrl = DEFAULT_V2_REGISTRY;
+		info.official = true;
+		info.indexName = DEFAULT_INDEX_NAME;
+	} else {
+		if (ns) {
+			info.remoteName = ns + '/' + name;
+		} else {
+			info.remoteName = name;
+		}
+		info.localName = index.name + '/' + info.remoteName;
+		info.canonicalName = info.localName;
+		info.indexUrl = `${index.scheme || 'https'}://${index.name}`;
+		info.official = false;
+		info.indexName = index.name;
+	}
+
+	return info as ParsedRepo;
+}
+
+function parseParams(header: string): HeaderParams {
```

Comment:
Your calling attention to this made me really rethink it.  We don't need such a complicated state-machine driven parser since we are only interested in two particular values.  I blew away this method and replaced it with some much simpler regex logic.